src/main/java/com/powsybl/balances_adjustment/util/ControlArea.java

```diff
@@ -0,0 +1,97 @@
+/**
+ * Copyright (c) 2021, RTE (http://www.rte-france.com)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.powsybl.balances_adjustment.util;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+import com.powsybl.cgmes.extensions.CgmesControlAreas;
+import com.powsybl.iidm.network.*;
+import org.jgrapht.Graph;
+import org.jgrapht.alg.connectivity.ConnectivityInspector;
+import org.jgrapht.graph.Pseudograph;
+
+/**
+ * @author Marcos de Miguel <demiguelm at aia.es>
+ */
+public class ControlArea implements NetworkArea {
+
+    private final Set<Object> terminalsAndBoundaries = new HashSet<>();
+    private final Set<String> voltageLevelIdsCache;
+
+    public ControlArea(Network network, String controlAreaId) {
+        CgmesControlAreas cgmesControlAreaMapping = network.getExtension(CgmesControlAreas.class);
+        terminalsAndBoundaries.addAll(cgmesControlAreaMapping.getCgmesControlArea(controlAreaId).getTerminals());
+        terminalsAndBoundaries.addAll(cgmesControlAreaMapping.getCgmesControlArea(controlAreaId).getBoundaries());
+
+        voltageLevelIdsCache = getVoltageLevelIds(network, terminalsAndBoundaries);
+    }
+
+    @Override
+    public double getNetPosition() {
+        return terminalsAndBoundaries.parallelStream().mapToDouble(o -> {
+            if (o instanceof Terminal) {
+                return getLeavingFlow((Terminal) o);
+            } else if (o instanceof Boundary) {
+                return -((Boundary) o).getP();
+            }
+            throw new AssertionError();
+        }).sum();
+    }
+
+    @Override
+    public Collection<String> getVoltageLevelIds() {
+        return Collections.unmodifiableSet(voltageLevelIdsCache);
+    }
+
+    private static Set<String> getVoltageLevelIds(Network network, Set<Object> terminalsAndBoundaries) {
+        return new ConnectivityInspector<>(createVoltageLevelGraph(network, terminalsAndBoundaries))
```

Comment:
Shouldn't you have kept the check that there's not more than one connected set?

Reply:
I would but `connectedSetOf(v)` returns only one connected sets (the biggest one containing v if there are several).

Reply:
yes, in fact there's only one connected set for the one vertex you're checking (the first element of `terminalsAndBoundaries`), but here the check we're talking about would be checking that all the connected sets corresponding to all vertices in `terminalsAndBoundaries` are the same, so checking that the size of the following set is 1:

```
  terminalsAndBoundaries.stream()
     .map(ControlArea::getVoltageLevel)
     .map(connectivityInspector::connectedSetOf)
     .collect(Collectors.toSet())
```